README
Credits:
    Mickael Leduque has lately contributed many features, such as
    faction editing, waypoint display and others.

    Alan Schmitt has been contributing a good chunk of code, especially
    the beginnings of a conversation editor.

    Thank you to Damon Law (damondesign.com) and Sylvan_Elf for the 
    neveredit icons and splash screen, respectively.

    Full credit should go to Torlack (torlack.com) for his great
    documentation on the mdl binary file format, and his nss script
    compiler (now maintained bythe openknights for linux and mac)
    that is used in neveredit.

    Neveredit contains some code from OpenGLContext (mainly the quaternion
    implementation), pyopengl.sf.net.

    I am now including the NWN Lexicon from http://nwnlexicon.com in
    the neveredit help. Thanks to all the contributors for putting together
    that great resource.

Building:
		Debian:
		If you are building on debian you will need to use PIP to install several python
		packages. LinearAlgebra, Pillow (Python Image Library Fork), and numpy.

Installation:
        If you have the Mac dmg distribution: drag the application to your
        applications folder. Double-click it there.

        If you have the Linux binary distribution: run 'neveredit' or 
        'neverscript' in the untarred/unzipped package 
       (whichever one you downloaded).

        If you're trying to run neveredit on another platform, or want
        to run neveredit from source for some other reason, or if the
        packaged version doesn't work for you at all:
        Check 
        http://neveredit.sourceforge.net/cgi-bin/moin.cgi/RunningNevereditFromSource
        for instructions as to how to satisfy the dependencies for running
        neveredit from source.

Instructions:
    So far, just a short note: I'm trying to emulate the camera controls
    in NWN in the map display, so try using up/down/left/right arrows
    as well as q,w,e,s. Thanks for jazzdman, these keys are now
    configurable in the prefrences.

Bugs and Comments:
    Please see http://neveredit.sf.net under 'For Users'.

Contact:
    sumpfork@users.sourceforge.net



ChangeLog
2020-09-20 11:12  dmjc

	updated array code to use modern numpy instead of num_array.

2006-01-18 11:30  sumpfork

	* __init__.py:

	bump version to 0.8.2b2

2006-01-10 11:00  sumpfork

	* ui/GLWindow.py (tags: v0_8_2_b2):

	fix missing import

2006-01-10 10:59  sumpfork

	* game/: Creature.py, Door.py, Item.py, Placeable.py, Trigger.py,
	  WayPoint.py (utags: v0_8_2_b2):

	fix missing quote for ints in warnings

2005-11-30 08:41  ledmicka

	* ui/SoundControl.py (tags: v0_8_2_b2):

	remove debug printing

2005-11-24 07:32  ledmicka

	* ui/PropWindow.py (tags: v0_8_2_b2):

	fix for tab/space problem

2005-11-24 07:30  ledmicka

	* ui/PropWindow.py:

	don't ignore the case of inexistant portrait - provide a default one

2005-11-21 06:06  ledmicka

	* ui/NeverEditMainApp.py (tags: v0_8_2_b2):

	fix some indentation errors (tabs v.s 4 spaces)

2005-11-18 18:21  sumpfork

	* Makefile (tags: v0_8_2_b2):

	fix cvsroot;show progress during upload

2005-11-17 20:24  jazzdman

	* resources/xrc/: PreferencesDialog.xrc, PreferencesDialog_xrc.py
	  (utags: v0_8_2_b2):

	Updated to allow user-defined nav keys for ModelWindow

2005-11-17 20:20  jazzdman

	* ui/NeverEditMainApp.py, ui/PreferencesDialog.py (tags:
	  v0_8_2_b2), ui/GLWindow.py, util/Preferences.py (tags:
	  v0_8_2_b2):

	Updated to allow user-defined nav keys for ModelWindow

2005-11-10 13:11  ledmicka

	* game/Module.py (tags: v0_8_2_b2, v0_8_2_b1):

	update the module area list after the addition of an ERF file

2005-11-10 06:11  ledmicka

	* ui/NeverEditMainApp.py (tags: v0_8_2_b1):

	fix a crash when re-selecting a tree item after an edition in the faction grid - we really have a problem with those tree selection events...

2005-11-10 06:09  ledmicka

	* ui/FactionGridWindow.py (tags: v0_8_2_b2, v0_8_2_b1):

	fix : the grid cells could not be edited

2005-11-09 19:41  sumpfork

	* ChangeLog (tags: v0_8_2_b2, v0_8_2_b1):

	update changelog

2005-11-09 19:40  sumpfork

	* __init__.py (tags: v0_8_2_b2, v0_8_2_b1):

	bump version to 0.8.2b1

2005-11-09 18:20  ledmicka

	* game/Module.py:

	use the basename of the resource included for ResRef instead of the full pathname
	use addResource instead of addRawResourceByName
	fix bug # 1352336

2005-11-09 18:18  ledmicka

	* file/ERFFile.py:

	getEntriesWithExtension did not return a list if only one element was found

2005-11-08 16:20  ledmicka

	* ui/PropWindow.py:

	use GetValue instead of GetStringSelection for ComboBox. This is a bug since control type change, and the user changes in this control weren't saved.

2005-11-08 16:18  ledmicka

	* ui/PropWindow.py:

	use good eventtype : having to hit enter after each edition is counter-intuitive

2005-11-08 16:16  ledmicka

	* ui/PropWindow.py:

	tab vs space fix

2005-11-08 14:22  ledmicka

	* ui/NeverEditMainApp.py:

	temporary fix for bug # 1291111 : not calling selectTreeItemById prevents triggering it - at the cost of autoselection of the object added

2005-11-08 14:17  ledmicka

	* ui/MapWindow.py:

	correction on a previous patch commited too hastily on MoveEvent.Clone method

2005-11-07 09:24  ledmicka

	* ui/MapWindow.py:

	replace the events Clone methods with ones that really create different objects, with distinct C++ part, so that segfault doesn't occur anymore when destroying both the original and the clone.

2005-10-30 17:13  ledmicka

	* ui/NeverEditMainApp.py:

	fix : the sound thread was not called at the good place, it only happened when one closed neveredit (not even when loading another module)

2005-10-26 23:00  sumpfork

	* ChangeLog:

	update changelog

2005-10-26 23:00  sumpfork

	* ui/SoundControl.py:

	initialize variable in error condition

2005-10-26 19:12  ledmicka

	* game/NeverData.py:

	adding some doc for NeverData class

2005-10-26 19:10  ledmicka

	* game/Trigger.py:

	adding a Neverdata subclass to handle triggers

2005-10-26 17:12  ledmicka

	* file/GFFFile.py:

	this should fix bug  1281419 on sf ('Music changes don't stick')

2005-10-25 18:11  ledmicka

	* ui/FactionGridWindow.py:

	disable removal of a faction - until the code is written

2005-10-25 18:07  ledmicka

	* game/Module.py, ui/NeverEditMainApp.py, ui/PropWindow.py:

	finalize faction integration into neveredit

2005-10-25 18:02  ledmicka

	* file/GFFFile.py:

	preparation for faction inclusion - allow to "flattenList" lists of GFFStructs subclasses instead of only lists of GFFSTructs

2005-10-25 17:59  ledmicka

	* ui/FactionGridWindow.py:

	adding a grid control to display and edit faction reactions

2005-10-25 17:56  ledmicka

	* game/Factions.py:

	adding a NeverData class to handle factions

2005-10-24 05:41  ledmicka

	* game/ResourceManager.py:

	add a correct non-emptiness test for Mod_CustomTlk

2005-10-17 10:09  ledmicka

	* game/Area.py, game/Palette.py, file/MDLFile.py, ui/MapWindow.py,
	  ui/NeverEditMainApp.py, ui/ToolPalette.py:

	add support for waypoint display (model), and addition (palette)

2005-10-17 09:59  ledmicka

	* game/WayPoint.py:

	add Module to read waypoint data from GFF

2005-10-17 09:42  ledmicka

	* game/Module.py:

	use logger.info instead of logger.error for informational message

2005-10-17 09:38  ledmicka

	* ui/GLWindow.py:

	add a call to glutInit() in ui.GLWindow - fixes a crash with a mouseover on an obect in a MapWindow (in mapWindow.DrawTextBox , with glutBitmapWidth with GLUT not initialized)

2005-10-16 08:59  sumpfork

	* ui/VarTableControl.py:

	removing erroneous/extraneous import

2005-10-14 10:14  ledmicka

	* game/ResourceManager.py:

	adding support for custom tlk reading

2005-10-14 09:56  ledmicka

	* game/Module.py, ui/PropWindow.py:

	adding support for VarTable - integration of the control into Neveredit

2005-10-13 17:27  ledmicka

	* ui/VarTableControl.py:

	adding a control for the VarTable property - basic support : adding a variable, modifying a value or a type. Limited to Int, String or Float

2005-10-13 14:58  ledmicka

	* game/Module.py:


ui/ToolPalette.py
"""A set of GUI classes showing blueprint palettes and a toolbar"""
import string

import wx

from neveredit.game.Palette import Palette
from neveredit.ui import WxUtils

#images via resourcepackage
from neveredit.resources.images import select_icon_png
from neveredit.resources.images import select_icon_sel_png
from neveredit.resources.images import paint_icon_png
from neveredit.resources.images import paint_icon_sel_png
from neveredit.resources.images import rotate_icon_png
from neveredit.resources.images import rotate_icon_sel_png

class PaletteWindow(wx.TreeCtrl):
    """A tree control representing a blueprint palette."""
    def __init__(self,parent,id):
        wx.TreeCtrl.__init__(self,parent,id,
                             style=wx.TR_DEFAULT_STYLE|wx.TR_HIDE_ROOT)
        self.AddRoot("Blueprint Palette")
        self.palette = None
        self.imagelist = wx.ImageList(16,26)
        self.SetImageList(self.imagelist)
        self.Bind(wx.EVT_TREE_ITEM_EXPANDING,self.itemExpanding)
        self.Bind(wx.EVT_TREE_SEL_CHANGING,self.selectionChanging)
        self.Bind(wx.EVT_TREE_ITEM_GETTOOLTIP,self.supplyToolTip)

    def fromPalette(self,palette):
        self.fromPaletteHelper(self.GetRootItem(),palette.getRoots())
        self.palette = palette

    def fromPaletteHelper(self,parentNode,childSpecs):
        for r in childSpecs:
            r.childrenReady = False
            node = self.AppendItem(parentNode,r.getName())
#            self.SetPyData(node,r)
            self.SetItemData(node,r)
            image = r.getImage()
            if image:
                image = image.crop((0,0,16,26))
                index = self.imagelist.Add(WxUtils.bitmapFromImage(image))
                self.SetItemImage(node,index)
            if r.getChildren():
                self.SetItemHasChildren(node,True)

    def itemExpanding(self,event):
        item = event.GetItem()
        data = self.GetItemData(item)
        if data and not data.childrenReady:
            self.fromPaletteHelper(item,data.getChildren())
            data.childrenReady = True

    def selectionChanging(self,event):
#        if event.GetItem().IsOk() and\
#           self.GetItemData(event.GetItem()) and\
#           not self.GetItemData(event.GetItem()).getBlueprint():
        if event.GetItem().IsOk() and\
            self.GetItemData(event.GetItem()) and\
            not self.GetItemData(event.GetItemData()).getBlueprint():
            event.Veto()

    def supplyToolTip(self,event):
        if event.GetItem().IsOk() and\
           self.GetItemData(event.GetItem()):
            bp = self.GetItemData(event.GetItem()).getBlueprint()
            if bp:
                event.SetToolTip(bp.getDescription())

    def get_standalone(cls, pname=None):
        class MyApp(wx.App):
            def OnInit(self):
                #if not pname:
                pname = 'creaturepalstd.itp'
                frame = wx.MiniFrame(None, -1, "Palette",
                                     wx.DefaultPosition, wx.Size(200,400))
                self.win = PaletteWindow(frame,-1)
                self.win.fromPalette(Palette.getStandardPalette('Creature'))
                frame.Show(True)
                self.SetTopWindow(frame)
                return True
        cls.app = MyApp(0)
        return cls.app.win
    get_standalone = classmethod(get_standalone)

    def start_standalone(cls):
        cls.app.MainLoop()
    start_standalone = classmethod(start_standalone)

TOOLSELECTIONEVENT = wx.NewEventType()

def EVT_TOOLSELECTION(window,function):
    '''notifies about the selection of a tool in the palette'''
    window.Connect(-1,-1,TOOLSELECTIONEVENT,function)

class ToolSelectionEvent(wx.PyCommandEvent):
    eventType = TOOLSELECTIONEVENT
    def __init__(self,windowID,tooltype):
        wx.PyCommandEvent.__init__(self,self.eventType,windowID)
        self.tooltype = tooltype
        self.data = None

    def setData(self,data):
        self.data = data

    def getData(self):
        return self.data

    def getToolType(self):
        return self.tooltype

    def Clone(self):
        self.__class__(self.GetId(),self.tooltype)

SELECTION_TOOL = wx.NewId()
ROTATE_TOOL = wx.NewId()
PAINT_TOOL = wx.NewId()

class ToolFrame(wx.MiniFrame):
    def __init__(self):
        import Image,WxUtils
        wx.MiniFrame.__init__(self,None,-1,"Tools",(805,25),(300,600))
        self.SetBackgroundColour('WHITE')
        self.CreateStatusBar()
        self.toolbar = self.CreateToolBar(wx.TB_FLAT | wx.NO_BORDER | wx.TB_HORIZONTAL)
        self.toolbar.SetBackgroundColour(wx.WHITE)
        self.toolbar.SetToolBitmapSize((26,24))
        self.selectId = SELECTION_TOOL
        self.toolbar.AddRadioTool(self.selectId,
                                       "Select/Move",
                                       select_icon_png.getBitmap(),
                                       select_icon_sel_png.getBitmap(),
                                       shortHelp=('Select Object'),
                                       longHelp='Select and Move objects on Map')
        self.paintId = PAINT_TOOL
        self.toolbar.AddRadioTool(self.paintId,
                                       "Paint",
                                       paint_icon_png.getBitmap(),
                                       paint_icon_sel_png.getBitmap(),
                                       shortHelp='Paint Objects',
                                       longHelp='Paint selected objects onto Map Display')
        self.rotateId = ROTATE_TOOL
        self.toolbar.AddRadioTool(self.rotateId,
                                       "Rotate",
                                       rotate_icon_png.getBitmap(),
                                       rotate_icon_sel_png.getBitmap(),
                                       shortHelp=('Rotate Object'),
                                       longHelp=('Rotate object shown on Map'))
        self.Bind(wx.EVT_TOOL,self.toolSelected)
        self.toolbar.AddSeparator()
        self.toolbar.Realize()
        self.toolIds = [self.selectId,self.paintId,self.rotateId]

        sublist = [ptype for ptype in Palette.PALETTE_TYPES
                   if ptype not in ['Sound','Encounter',
                                    'Trigger','Store','Item']]
        self.stdPalettes = dict(zip(sublist,
                                    [Palette.getStandardPalette(ptype)
                                     for ptype in sublist]))
        self.notebook = wx.Notebook(self,-1,style=wx.NB_LEFT)
        for type,palette in self.stdPalettes.iteritems():
            pw = PaletteWindow(self.notebook,-1)
            pw.fromPalette(palette)
            self.notebook.AddPage(pw,type)
            pw.Bind(wx.EVT_TREE_SEL_CHANGED,self.treeItemSelected)

        self.toggleToolOn(self.selectId)
        self.lastPaletteSelection = None

    def getActivePaletteWindow(self):
        return self.notebook.GetPage(self.notebook.GetSelection())

    def toggleToolOn(self,id):
        for tid in self.toolIds:
            self.toolbar.ToggleTool(tid,tid==id)
        newEvent = ToolSelectionEvent(self.GetId(),id)
        if id != self.paintId:
            self.lastPaletteSelection = self.getActivePaletteWindow().GetSelection()
            self.getActivePaletteWindow().Unselect()
        else:
            if self.lastPaletteSelection:
                self.getActivePaletteWindow().SelectItem(self.lastPaletteSelection)
                self.lastPaletteSelection = None
            bp = self.getSelectedBlueprint()
            if bp:
                newEvent.setData(bp.toInstance())
#        self.GetEventHandler().AddPendingEvent(newEvent)

    def treeItemSelected(self,event):
        if self.getActivePaletteWindow().GetItemData(event.GetItem()):
#        if self.getActivePaletteWindow().GetItemData(event.GetItemData()):
            self.toggleToolOn(self.paintId)
        event.Skip()

    def getSelectedBlueprint(self):
        palette = self.getActivePaletteWindow()
        data = palette.GetItemData(palette.GetSelection())
        if data:
            return data.getBlueprint()

    def toolSelected(self,event):
        self.toggleToolOn(event.GetId())
        event.Skip()

if __name__ == "__main__":
    class MyApp(wx.App):
        def OnInit(self):
            wx.InitAllImageHandlers()
            f = ToolFrame()
            f.Show(True)
            self.SetTopWindow(f)
            return True
    app = MyApp(0)
    app.MainLoop()


